src/app/view/editor-main-view/data-views/trainrunsections.view.ts

```diff
@@ -2147,6 +2348,133 @@ export class TrainrunSectionsView {
     return retPath;
   }
 
+  /**
+   * Detect collapsed node chains and calculate direct visualization path
+   *
+   * A collapsed chain occurs when intermediate nodes are hidden from view.
+   * Example: A-B*-C*-D where B* and C* are collapsed -> create direct path A→D
+   *
+   * Algorithm:
+   * 1. Walk backward from current section to find chain start (non-collapsed node)
+   * 2. Walk forward from current section to find chain end (non-collapsed node)
+   * 3. If start != end, calculate direct path using legacy routing system
+   *
+   * @param currentSection The section that might be part of a collapsed chain
+   * @returns Direct path Vec2D[] if chain detected, null otherwise
+   */
+  private getCollapsedChainPath(currentSection: TrainrunSection): Vec2D[] | null {
+    const trainrunId = currentSection.getTrainrunId();
+
+    const startSection = this.findChainBoundary(currentSection, trainrunId, "backward");
+    const endSection = this.findChainBoundary(currentSection, trainrunId, "forward");
+
+    if (startSection.getId() !== endSection.getId()) {
+      const startNode = startSection.getSourceNode();
+      const endNode = endSection.getTargetNode();
+
+      return TrainrunSectionsView.calculateDirectPath(
+        startNode,
+        endNode,
+        [startSection, endSection],
+        currentSection,
+      );
+    }
+
+    return null;
+  }
+
+  /**
+   * Updates the TrainrunSection's path if it's part of a collapsed chain.
+   * This ensures the visual rendering uses the correct collapsed path instead of the direct path.
+   */
+  private updateTrainrunSectionPathForCollapsedChain(section: TrainrunSection): void {
+    const collapsedPath = this.getCollapsedChainPath(section);
+    if (collapsedPath) {
+      (section as any).pathVec2D = collapsedPath;
+
+      const sourcePort = section.getSourceNode().getPort(section.getSourcePortId());
+      if (sourcePort) {
+        (section as any).path.textPositions =
+          SimpleTrainrunSectionRouter.placeTextOnTrainrunSection(collapsedPath, sourcePort);
+      }
+    }
+  }
+
+  /**
+   * Find the boundary of a collapsed chain by walking in the specified direction
+   */
+  private findChainBoundary(
+    startSection: TrainrunSection,
+    trainrunId: number,
+    direction: "forward" | "backward",
+  ): TrainrunSection {
+    let currentSection = startSection;
+
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      const node =
+        direction === "backward" ? currentSection.getSourceNode() : currentSection.getTargetNode();
+
+      if (!node.getIsCollapsed()) {
+        break;
+      }
+
+      const connectedSections = this.getConnectedTrainrunSections(node, trainrunId).filter(
+        (ts) => ts.getId() !== currentSection.getId(),
+      );
+
+      if (connectedSections.length === 1) {
+        currentSection = connectedSections[0];
+      } else {
+        break;
+      }
```

Comment:
If I understand correctly, the goal here is to break if there is no connection. Then, it would be clearer to state this in the if condition:
```suggestion
      if (connectedSections.length === 0) {
        break;
      } else {
        currentSection = connectedSections[0];
      }
```

Also, is it possible to have `connectedSections.length > 1` ? is it normal to break right now if it is the case ?